src/Handler/Auth.hs

```diff
@@ -159,3 +159,66 @@ getSignoutR :: Handler Html
 getSignoutR = do
   deleteLoginData
   redirect HomeR
+
+getForgotR :: Handler Html
+getForgotR = do
+  redirectIfLoggedIn HomeR
+  (forgotFormWidget, _) <- generateFormPost forgotForm
+  renderForgot forgotFormWidget []
+
+postForgotR :: Handler Html
+postForgotR = do
+  redirectIfLoggedIn HomeR
+  ((result, widget), _) <- runFormPost forgotForm
+  case result of
+    FormSuccess email -> do
+      maybeUP <- runDB $ getUserByEmail email
+      case maybeUP of
+        (Just (Entity userKey _)) -> do
+          (Entity _ (Reset token _ _)) <- runDB $ do
+            deleteExistingResets userKey
+            createReset userKey
+          -- send link in email
+          $logInfo $ tokenText token
+          renderNotice "Success" ["Please check your email to reset your password."]
+        Nothing -> do
+          renderForgot widget ["This user does not exist."]
+    _ -> renderForgot widget []
+
+getResetR :: Token -> Handler Html
+getResetR token = do
+  redirectIfLoggedIn HomeR
+  maybeUser <- runDB $ do
+    deleteOldResets
```

Comment:
I like the idea of purging old resets! There's a potential problem with this approach:

If `getUserByResetToken` threw an exception for some reason, the purge would be rolled back and I don't think that's desired in this case.

This is a minor complaint! I don't think this needs to be changed to make the PR acceptable at all.

Reply:
Yeah true, I hadn't thought of that.  Splitting to two transactions is easy enough.